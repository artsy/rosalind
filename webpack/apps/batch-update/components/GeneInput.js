import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { colors } from './Layout'

class GeneInput extends React.Component {
  constructor (props) {
    super(props)
    this.handleChange = this.handleChange.bind(this)
  }

  handleChange (e) {
    const { onChangeValue, name } = this.props
    const value = e.target.value
    onChangeValue({ name, value })
  }

  render () {
    const { name, value } = this.props
    return (
      <Square value={value}>
        <Name value={value}>
          { value === 0 ? <del>{name}</del> : name }
          <Bar value={value} />
        </Name>
        <Value
          type='number'
          min='0'
          max='100'
          step='10'
          placeholder='Multiple'
          value={value === null ? '' : value}
          onChange={this.handleChange}
        />
      </Square>
    )
  }
}

GeneInput.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.number,
  onChangeValue: PropTypes.func
}

const Square = styled.div`
  ${props => {
    if (parseInt(props.value) === 0) {
      return `
        border: solid 1px ${colors.red};
      `
    } else if (props.value) {
      return `
        border: solid 1px ${colors.purple};
      `
    } else {
      return `
        border: solid 1px ${colors.grayLighter};
      `
    }
  }}
  width: 140px;
  height: 140px;
  margin-right: 20px;
  margin-bottom: 20px;
  display: flex;
  flex-direction: column;
`
Square.displayName = 'Square'

const Name = styled.div`
  ${props => {
    if (parseInt(props.value) === 0) {
      return `
        background: ${colors.redLightest};
        border-bottom: solid 1px ${colors.red};
      `
    } else if (props.value) {
      return `
        background: ${colors.purpleLightest};
        border-bottom: solid 1px ${colors.purple};
      `
    } else {
      return `
        background: ${colors.grayLightest};
        border-bottom: solid 1px ${colors.grayLighter};
      `
    }
  }}
  flex-basis: 100px;
  padding: 0.5em 1em;
  overflow: hidden;
  position: relative;
`
Name.displayName = 'Name'

const Value = styled.input`
  flex-basis: 40px;
  border-color: transparent;
  padding: 1em;
  &:focus {
    outline: solid 1px ${colors.gray};
  }
`
Value.displayName = 'Value'

const Bar = styled.div`
  position: absolute;
  left: 0;
  bottom: 0;
  width: 100%;
  background: hsla(261, 100%, 85%, 0.15);
  transition: height 0.15s;
  ${props => {
    return `
      height: ${props.value}%;
    `
  }}
`
Bar.displayName = 'Bar'

export default GeneInput
