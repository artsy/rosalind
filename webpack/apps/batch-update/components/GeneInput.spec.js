import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import GeneInput from './GeneInput'

it('renders a null gene correctly', () => {
  const rendered = renderer.create(<GeneInput name='Kawaii' value={null} />)
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders a zero-valued gene correctly', () => {
  const rendered = renderer.create(<GeneInput name='Kawaii' value={0} />)
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders a positive-valued gene correctly', () => {
  const rendered = renderer.create(<GeneInput name='Kawaii' value={70} />)
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

it('accepts changes to the gene value via its <input> element', () => {
  const changeHandler = jest.fn()
  const wrapper = mount(<GeneInput name='Kawaii' value={null} onChangeValue={changeHandler} />)
  const mockEvent = { target: { value: '70' } }
  wrapper.find('input').simulate('change', mockEvent)
  expect(changeHandler.mock.calls.length).toEqual(1)
  expect(changeHandler.mock.calls[0][0]).toEqual({name: 'Kawaii', value: '70'})
})
