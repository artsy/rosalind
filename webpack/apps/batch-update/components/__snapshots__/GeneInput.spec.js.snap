// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a null gene correctly 1`] = `
<div
  className="dDtKEb"
  value={null}
>
  <div
    className="iXxpJJ"
    value={null}
  >
    Kawaii
    <div
      className="gjNVml"
      value={null}
    />
  </div>
  <input
    className="kjtXRk"
    max="100"
    min="0"
    onChange={[Function]}
    placeholder="Multiple"
    step="10"
    type="number"
    value=""
  />
</div>
`;

exports[`renders a positive-valued gene correctly 1`] = `
<div
  className="eJIUEQ"
  value={70}
>
  <div
    className="eaagZv"
    value={70}
  >
    Kawaii
    <div
      className="kHxQBi"
      value={70}
    />
  </div>
  <input
    className="kjtXRk"
    max="100"
    min="0"
    onChange={[Function]}
    placeholder="Multiple"
    step="10"
    type="number"
    value={70}
  />
</div>
`;

exports[`renders a zero-valued gene correctly 1`] = `
<div
  className="inyXqK"
  value={0}
>
  <div
    className="cXizAY"
    value={0}
  >
    <del>
      Kawaii
    </del>
    <div
      className="kdkQr"
      value={0}
    />
  </div>
  <input
    className="kjtXRk"
    max="100"
    min="0"
    onChange={[Function]}
    placeholder="Multiple"
    step="10"
    type="number"
    value={0}
  />
</div>
`;
